README.md
# hydroGOF
[![Research software impact](http://depsy.org/api/package/cran/hydroGOF/badge.svg)](http://depsy.org/package/r/hydroGOF) [![CRAN](http://www.r-pkg.org/badges/version/hydroGOF)](https://cran.r-project.org/package=hydroGOF) [![License](https://img.shields.io/badge/license-GPL%20%28%3E=%203%29-lightgrey.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html) [![monthly](http://cranlogs.r-pkg.org/badges/hydroGOF)](https://www.rpackages.io/package/hydroGOF) [![total](http://cranlogs.r-pkg.org/badges/grand-total/hydroGOF)](https://www.rpackages.io/package/hydroGOF) [![Build Status](https://travis-ci.org/hzambran/hydroGOF.svg?branch=master)](https://travis-ci.org/hzambran/hydroGOF) [![dependencies](https://tinyverse.netlify.com/badge/hydroGOF)](https://CRAN.R-project.org/package=hydroGOF)

hydroGOF is an R package that provides S3 functions implementing both statistical and graphical goodness-of-fit measures between observed and simulated values, mainly oriented to be used during the calibration, validation, and application of hydrological models.
Missing values in observed and/or simulated values can removed before the computations.
Bugs / comments / questions / collaboration of any kind are very welcomed. 



## Installation
Installing the latest stable version from [CRAN](https://CRAN.R-project.org/package=hydroGOF):
```{r}
install.packages("hydroGOF")
```

Alternatively, you can also try the under-development version from [Github](https://github.com/hzambran/hydroGOF):
```{r}
if (!require(devtools)) install.packages("devtools")
library(devtools)
install_github("hzambran/hydroGOF")
```


## Reporting bugs, requesting new features

If you find an error in some function, or want to report a typo in the documentation, or to request a new feature (and wish it be implemented :) you can do it [here](https://github.com/hzambran/hydroGOF/issues)


## Citation 
```{r}
citation("hydroGOF")
```

To cite hydroGOF in publications use:

>  Mauricio Zambrano-Bigiarini. hydroGOF: Goodness-of-fit functions for comparison of simulated and observed hydrological time series. R package version 0.4-0. URL https://github.com/hzambran/hydroGOF. DOI:10.5281/zenodo.839854.


A BibTeX entry for LaTeX users is

>  @Manual{hydroGOF,  
>    title = {hydroGOF: Goodness-of-fit functions for comparison of simulated and observed hydrological time series},  
>    author = {{Mauricio Zambrano-Bigiarini}},  
>    note = {R package version 0.4-0},  
>    url = {https://github.com/hzambran/hydroGOF},  
>    doi = {DOI:10.5281/zenodo.839854},  
>  }


## Vignette 
[Here](https://cran.r-project.org/web/packages/hydroGOF/vignettes/hydroGOF_Vignette.pdf) you can find an introductory vignette showing the use of several hydroGOF functions.


## Related Material 

* *R: a statistical environment for hydrological analysis* (**EGU-2010**)  [abstract](http://meetingorganizer.copernicus.org/EGU2010/EGU2010-13008.pdf), [poster](http://www.slideshare.net/hzambran/egu2010-ra-statisticalenvironmentfordoinghydrologicalanalysis-9095709).

* *Comparing Goodness-of-fit Measures for Calibration of Models Focused on Extreme Events* (**EGU-2012**) [abstract](http://meetingorganizer.copernicus.org/EGU2012/EGU2012-11549-1.pdf), [poster](http://www.slideshare.net/hzambran/egu2012-11549go-fsforextremeevents4web).

* *Using R for analysing spatio-temporal datasets: a satellite-based precipitation case study* (**EGU-2017**) [abstract](http://meetingorganizer.copernicus.org/EGU2017/EGU2017-18343.pdf), [poster](https://doi.org/10.5281/zenodo.570145).



## See Also 

* [hydroTSM: Time series management, analysis and interpolation for hydrological modelling](https://github.com/hzambran/hydroTSM).

* [hydroPSO: Model-independent Particle Swarm Optimisation (PSO) for environmental/hydrological models](https://github.com/hzambran/hydroPSO).




NEWS
NEWS/ChangeLog for hydroGOF
--------------------------
0.5-0  Under-development
        o new function 'KGElf', which computes the Kling-Gupta efficiency focused on low flows proposed by García et al. (2017).
        o new function 'sKGE', which computes the Split Kling-Gupta efficiency, which was proposed by Fowler et al. (2018) for calibration in a drying climate.
        o new function 'KGEnp', which computes the non-parametric Kling-Gupta efficiency proposed by Pool et al. (2018).
        o new function 'dr', which computes the Refined Index of Agreement proposed by Willmott et al. (2012).
        o 'KGE'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of KGE (e.g., log, sqrt, 1/x)   
        o 'mNSE'    : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of mNSE (e.g., log, sqrt, 1/x) 
        o 'rNSE'    : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of rNSE (e.g., log, sqrt, 1/x)                                                                                           
        o 'd'       : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of d (e.g., log, sqrt, 1/x)                                                                                         
        o 'rd'      : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of rd (e.g., log, sqrt, 1/x)  
        o 'VE'      : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of VE (e.g., log, sqrt, 1/x)  
        o 'cp'      : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of cp (e.g., log, sqrt, 1/x)  
        o 'mae'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of mae (e.g., log, sqrt, 1/x)  
        o 'me'      : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of me (e.g., log, sqrt, 1/x)
        o 'mse'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of mse (e.g., log, sqrt, 1/x) 
        o 'nrmse'   : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of nrmse (e.g., log, sqrt, 1/x)     
        o 'pbias'   : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of pbias (e.g., log, sqrt, 1/x)  
        o 'pbiasfdc': three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of pbiasfdc (e.g., log, sqrt, 1/x)  
        o 'rSD'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of rSD (e.g., log, sqrt, 1/x)  
        o 'rsr'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of rsr (e.g., log, sqrt, 1/x)  
        o 'br2'     : three new arguments (fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of br2 (e.g., log, sqrt, 1/x)  
        o 'br2'     : new argument 'dec' to allow the user to specify the number of decimals shown in the output object. Default value set to 1 for compatibility with previous versions. Thanks to Ken Newman !
        o 'br2'     : R2 is now correctly computed as 1 - SSres/SStot. Before it was computed as rPearson^2, whcih was only correct for linear models. Thanks to eugenioLR (https://github.com/hzambran/hydroGOF/issues/16#issue-1736556320)
        o 'NSE'     : possible values of 'epsilon' argument were changed to c('Pushpalatha2012', 'otherFactor', 'otherValue'), to give the user more flexibility in the computation of the constant value added to 'sim' and 'obs' when using transformed values to compute the NSE (e.g., when focusing on low flows)
        o 'gof'     : four new arguments (start.month, fun, epsilon, epsilon.value), added to allow pre-processing simulated and observed values before computation of all goodness-of-fit.                                                                                                                                                                      

0.4-0  11-Mar-2020
        o Package tested against R Under development (unstable) (2020-03-10 r77920) -- "Unsuffered Consequences", following an imperative request made by CRAN.
        o Citation file changed, following CRAN comments.
        o Vignette on Goodness-of-fit Measures to Compare Observed and Simulated Values was moved from Sweave to Knitr.
        o New references were added for 'KGE' (Santos et. al, 2018; Knoben et al., 2019; Mizukami et al., 2019)
        o New reference was added for 'me' (Hill et al., 2006) Thanks to Erli Pinto dos Santos !.
        o 'br2'     : new argument 'use.abs=FALSE', to allow the user to use 'abs(b)' as condition to decide whether using abs(b)*r2 or [1/abs(b)]*r2 in equation (5) in Krausse et al. (2005). Thanks to Ellie White !



0.3-9  07-Ago-2017
        o repository management moved from SVN to GIT, including rforge.
        o citation with DOI is now possible (and new CITATION file).

R/dr.R
########################################
# 'dr': Refined Index of Agreement     #
########################################
# December 18th, 2008;  06-Sep-09      #
# 28-Feb-2016                          #
# Updates: 20-Jul-2022 ; 29-Jul-2022   #
########################################
# 1) Willmott, C.J., Robeson, S.M. and Matsuura, K. (2012). A refined index of model performance. International Journal of climatology, 32(13), pp.2088-2094. doi:10.1002/joc.2419.
# 2) Willmott, C.J., Robeson, S.M., Matsuura, K. and Ficklin, D.L. (2015). Assessment of three dimensionless measures of model performance. Environmental Modelling & Software, 73, pp.167-174. doi:10.1016/j.envsoft.2015.08.012
# 3) Willmott, C.J. 1981. On the validation of models. Physical Geography, 2, 184-194
# 4) Willmott, C. J. (1984). "On the evaluation of model performance in physical geography." Spatial Statistics and Models, G. L. Gaile and C. J. Willmott, eds., 443-460.
# 5) Legates, D. R., and G. J. McCabe Jr. (1999), Evaluating the Use of "Goodness-of-Fit" Measures in Hydrologic and Hydroclimatic Model Validation, Water Resour. Res., 35(1), 233-241. 


# In contrast to the Index of Agreement (Willmott et al., 1984), which ranges from 0 to 1.0,
# the Refined Index of Agreement is bounded by -1.0 and 1.0. The closer to 1 the better 
# the performance of the model.
# In geneeral, dr is  "more rationally related to model accuracy than are other existing
# indices" (MSE, NSE, d, rd) (Willmott et a., 2012; Willmott et al., 2015)

# 'obs'   : numeric 'data.frame', 'matrix' or 'vector' with observed values
# 'sim'   : numeric 'data.frame', 'matrix' or 'vector' with simulated values

# 'Result': Index of Agreement between 'sim' and 'obs'

dr <-function(sim, obs, ...) UseMethod("dr")

################################################################################
# Author: Mauricio Zambrano-Bigiarini                                          #
################################################################################
# Started: 20-Jul-2022                                                         #
# Updates: 29-Jul-2022                                                         #
################################################################################
dr.default <- function(sim, obs, na.rm=TRUE,
                      fun=NULL, ...,
                      epsilon.type=c("none", "Pushpalatha2012", "otherFactor", "otherValue"), 
                      epsilon.value=NA){ 

     if ( is.na(match(class(sim), c("integer", "numeric", "ts", "zoo"))) |
          is.na(match(class(obs), c("integer", "numeric", "ts", "zoo")))
     ) stop("Invalid argument type: 'sim' & 'obs' have to be of class: c('integer', 'numeric', 'ts', 'zoo')")

     epsilon.type <- match.arg(epsilon.type)

     # index of those elements that are present both in 'x' and 'y' (NON- NA values)
     vi <- valindex(sim, obs)
     
     if (length(vi) > 0) {	 
       # Filtering 'obs' and 'sim', selecting only those pairs of elements 
       # that are present both in 'x' and 'y' (NON- NA values)
       obs <- obs[vi]
       sim <- sim[vi]

       if (!is.null(fun)) {
         fun1 <- match.fun(fun)
         new  <- preproc(sim=sim, obs=obs, fun=fun1, ..., 
                         epsilon.type=epsilon.type, epsilon.value=epsilon.value)
         sim  <- new[["sim"]]
         obs  <- new[["obs"]]
       } # IF end
     
       # the next two lines are required for avoiding an strange behaviour 
       # of the difference function when sim and obs are time series.
       if ( !is.na(match(class(sim), c("ts", "zoo"))) ) sim <- as.numeric(sim)
       if ( !is.na(match(class(obs), c("ts", "zoo"))) ) obs <- as.numeric(obs)
     
       # Mean of the observed values
       Om <- mean(obs, na.rm=na.rm)

       # Constant 'c' value
       c <- 2

       # Components of the denominator
       A <- sum( abs(sim - obs) )
       B <- c*sum( abs(obs - Om) )
     
       if (A <= B) {      
         dr <- 1 - A / B     
       } else dr <- 1 - B / A 
     } else {
         dr <- NA
         warning("There are no pairs of 'sim' and 'obs' without missing values !")
       } # ELSE end
     
     return(dr) 
     
} # 'dr.default' end


################################################################################
# Author: Mauricio Zambrano-Bigiarini                                          #
################################################################################
# Started: 20-Jul-2022                                                         #
# Updates: 29-Jul-2022                                                         #
################################################################################
dr.matrix <- function(sim, obs, na.rm=TRUE,
                     fun=NULL, ...,
                     epsilon.type=c("none", "Pushpalatha2012", "otherFactor", "otherValue"), 
                     epsilon.value=NA){ 
 
 # Checking that 'sim' and 'obs' have the same dimensions
 if ( all.equal(dim(sim), dim(obs)) != TRUE )
    stop( paste("Invalid argument: dim(sim) != dim(obs) ( [", 
          paste(dim(sim), collapse=" "), "] != [", 
          paste(dim(obs), collapse=" "), "] )", sep="") )

 dr <- rep(NA, ncol(obs))       
          
 dr <- sapply(1:ncol(obs), function(i,x,y) { 
             dr[i] <- dr.default( x[,i], y[,i], na.rm=na.rm, ... )
             }, x=sim, y=obs, na.rm=na.rm, fun=fun, ..., 
             epsilon.type=epsilon.type, epsilon.value=epsilon.value)    
                     
  names(dr) <- colnames(obs)
  return(dr)
     
} # 'dr.matrix' end


dr.data.frame <- function(sim, obs, na.rm=TRUE,
                         fun=NULL, ...,
                         epsilon.type=c("none", "Pushpalatha2012", "otherFactor", "otherValue"), 
                         epsilon.value=NA){  
 
  sim <- as.matrix(sim)
  obs <- as.matrix(obs)
   
  dr.matrix(sim, obs, na.rm=na.rm, fun=fun, ..., 
            epsilon.type=epsilon.type, epsilon.value=epsilon.value)    
     
} # 'dr.data.frame' end


################################################################################
# Author: Mauricio Zambrano-Bigiarini                                          #
################################################################################
# Started: 20-Jul-2022                                                         #
# Updates: 29-Jul-2022                                                         #
################################################################################
dr.zoo <- function(sim, obs, na.rm=TRUE,
                  fun=NULL, ...,
                  epsilon.type=c("none", "Pushpalatha2012", "otherFactor", "otherValue"), 
                  epsilon.value=NA){ 
    
    sim <- zoo::coredata(sim)
    if (is.zoo(obs)) obs <- zoo::coredata(obs)
    
    if (is.matrix(sim) | is.data.frame(sim)) {
       dr.matrix(sim, obs, na.rm=na.rm, fun=fun, ..., 
                epsilon.type=epsilon.type, epsilon.value=epsilon.value)    
    } else NextMethod(sim, obs, na.rm=na.rm, fun=fun, ..., 
                      epsilon.type=epsilon.type, epsilon.value=epsilon.value)    
     
  } # 'dr.zoo' end
